### PR TITLE
Speed up software CRC-32 computation by a factor of 1.5 to 3.

### PR DESCRIPTION
**Pull request by nmoinvaz**
Use the interleaved method of Kadatch and Jenkins in order to make
use of pipelined instructions through multiple ALUs in a single
core. This also speeds up and simplifies the combination of CRCs,
and updates the functions to pre-calculate and use an operator for
CRC combination.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
This PR is still a work in progress at least it passes `gtest_zlib` on my machine. I kept both CRC32 methods in so I could benchmark the improvement:

```
2022-05-08T13:09:05-07:00
Running C:\Users\nathan\Source\zlib-ng\build\test\benchmarks\Release\benchmark_zlib.exe
Run on (12 X 3192 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 12288 KiB (x1)
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
crc32/gf2_byfour/1                2.75 ns         2.73 ns    263529412
crc32/gf2_byfour/8                7.04 ns         7.15 ns     89600000
crc32/gf2_byfour/64               49.0 ns         48.8 ns     11200000
crc32/gf2_byfour/512               418 ns          425 ns      1544828
crc32/gf2_byfour/4096             3299 ns         3278 ns       224000
crc32/gf2_byfour/32768           25645 ns        25495 ns        26353
crc32/gf2_byfour/262144         216420 ns       213109 ns         3446
crc32/gf2_byfour/2097152       1853120 ns      1848493 ns          448
crc32/gf2_byfour/4194304       3453858 ns      3447770 ns          213
crc32/braid/1                     3.03 ns         3.00 ns    213333333
crc32/braid/8                     15.8 ns         15.3 ns     40727273
crc32/braid/64                    53.1 ns         51.6 ns     10000000
crc32/braid/512                    187 ns          184 ns      3733333
crc32/braid/4096                  1224 ns         1196 ns       640000
crc32/braid/32768                 9314 ns         9417 ns        74667
crc32/braid/262144               74811 ns        73242 ns         8960
crc32/braid/2097152             607181 ns       610352 ns          896
crc32/braid/4194304            1259157 ns      1255020 ns          498
crc32/pclmulqdq/1                 4.89 ns         4.88 ns    112000000
crc32/pclmulqdq/8                 7.23 ns         7.15 ns     89600000
crc32/pclmulqdq/64                21.1 ns         21.3 ns     34461538
crc32/pclmulqdq/512               42.8 ns         42.6 ns     17230769
crc32/pclmulqdq/4096               187 ns          186 ns      3200000
crc32/pclmulqdq/32768             1398 ns         1395 ns       448000
crc32/pclmulqdq/262144           12878 ns        12765 ns        74667
crc32/pclmulqdq/2097152          94831 ns        95215 ns         6400
crc32/pclmulqdq/4194304         197139 ns       199507 ns         3446
```

**Comment by KungFuJesus** [comment]
Awesome, was hoping somebody would port these changes in soon :).  This will benefit potato CPUs that lack carryless multiply (which actually includes all raspberry pis since they lack the cryptographic instructions).

**Comment by nmoinvaz** [comment]
@Dead2 right now I have both crc methods in the tree. Would you rather I build on top of this commit and then add a commit to remove the old method? Or do you prefer I just nuke the old method in the same commit?

**Comment by Dead2** [comment]
Another alternative that is commonly used for example for the linux kernel is to have a prepare-commit that just adds all the new files without actually hooking anything up yet, and then the next commit is all about doing the actual replacement, and last a commit for removing any old files that are now orphaned. That makes it relatively easy to review wholesale replacements of a subsystem while always remaining fully buildable in each commit (important for bisect).

If not that, then do it in a single commit since there is a lot of renaming of the old stuff already in the first commit, and that just adds lots of extra changes to review that would go away in the next commit.

**Comment by nmoinvaz** [comment]
I think I will just save this state of the PR as a separate branch in my fork. https://github.com/nmoinvaz/zlib-ng/tree/backup/crc32-braid. Then I will continue with this PR and do it as a single commit.

**Comment by nmoinvaz** [comment]
Looks like now only most of the pkgchecks fail which is expected.

**Comment by nmoinvaz** [comment]
I ran some tests against #1228 on top of this PR and it seems like we are going to have to merge in this PR before we can update #1228 and the abi files and commits.

**Comment by nmoinvaz** [comment]
I tested it here in a separate branch https://github.com/nmoinvaz/zlib-ng/runs/6401946125?check_suite_focus=true. I had to also update the commit for madler repo to latest in `develop` because of shared object generation issue in configure script.

**Comment by nmoinvaz** [comment]
Rebased.

File: arch/x86/crc32_fold_pclmulqdq.c [hunk 1]
```diff
@@ -597,7 +597,7 @@ uint32_t crc32_pclmulqdq(uint32_t crc32, const unsigned char* buf, uint64_t len)
     /* For lens < 64, crc32_byfour method is faster. The CRC32 instruction for
```
Comment (Dead2) [inline arch/x86/crc32_fold_pclmulqdq.c]: I think this comments needs updating to crc32_braid

Comment (nmoinvaz) [reply arch/x86/crc32_fold_pclmulqdq.c]: Fixed.

**Comment by Dead2** [comment]
# x86-64
### baseline d79984b5bcaccab15e6cd13d7d1edea32ac36977
```
 Tool: minigzip   Levels: 1-9
 Runs: 15         Trim worst: 8
 
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.052/1.072/1.079/0.009        0.506/0.512/0.518/0.005       94,127,290
 2     35.518%      1.990/2.008/2.021/0.012        0.540/0.546/0.554/0.005       75,282,961
 3     33.882%      2.503/2.528/2.543/0.013        0.518/0.524/0.528/0.003       71,816,478
 4     33.174%      2.835/2.879/2.902/0.022        0.503/0.509/0.512/0.003       70,315,668
 5     32.660%      3.112/3.172/3.206/0.031        0.497/0.504/0.507/0.003       69,225,542
 6     32.508%      3.724/3.772/3.817/0.041        0.494/0.501/0.503/0.003       68,902,222
 7     32.255%      5.130/5.188/5.234/0.036        0.491/0.497/0.500/0.003       68,366,800
 8     32.167%      7.809/7.874/7.934/0.046        0.490/0.499/0.504/0.005       68,180,776
 9     31.887%      8.479/8.538/8.590/0.038        0.490/0.496/0.499/0.003       67,586,442

 avg1  34.273%                        4.114                          0.510
 tot                                259.211                         32.121      653,804,179

   text    data     bss     dec     hex filename
 135364    1392      48  136804   21664 libz-ng.so.2
 ```
 
 ### PR #1268 90c51c8f4d9f26cca159bcdae2416907cb766cbd
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.065/1.075/1.082/0.006        0.505/0.509/0.511/0.002       94,127,290
 2     35.518%      1.995/2.006/2.011/0.006        0.534/0.540/0.544/0.004       75,282,961
 3     33.882%      2.521/2.534/2.549/0.010        0.511/0.518/0.522/0.005       71,816,478
 4     33.174%      2.874/2.896/2.911/0.014        0.494/0.500/0.507/0.005       70,315,668
 5     32.660%      3.168/3.205/3.219/0.019        0.485/0.494/0.500/0.005       69,225,542
 6     32.508%      3.722/3.752/3.776/0.026        0.486/0.491/0.496/0.004       68,902,222
 7     32.255%      5.135/5.157/5.175/0.017        0.482/0.487/0.490/0.003       68,366,800
 8     32.167%      7.872/7.907/7.942/0.026        0.485/0.494/0.499/0.005       68,180,776
 9     31.887%      8.495/8.521/8.573/0.028        0.477/0.487/0.492/0.005       67,586,442

 avg1  34.273%                        4.117                          0.502
 tot                                259.371                         31.644      653,804,179

   text    data     bss     dec     hex filename
 133180    1392      48  134620   20ddc libz-ng.so.2
 ```
Looks like average ~1.5% faster inflate, not sure why, possibly it means it falls back to crc32-braid from the crc32-pclmulqdq for short lengths a lot.

**Review by Dead2 — APPROVED** [review]

**Comment by nmoinvaz** [comment]
That sounds probable. I expect it would show even more performance improvement if comparing without optimization.

**Comment by nmoinvaz** [comment]
Am I reading that benchmark right that there is a code size reduction? That is surprising. I thought perhaps the new tables were larger.

**Comment by Dead2** [comment]
Yes, I forgot to mention that, but I actually had to double-check that myself. The same reduction can be seen on Aarch64 too, but the benchmarks were much closer there.

**Pull request merged**